rbac/management/principal/it_service.py

```diff
@@ -0,0 +1,347 @@
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+"""Class to manage interactions with the IT service accounts service."""
+import logging
+import time
+import uuid
+
+import requests
+from django.conf import settings
+from django.db.models import Q
+from management.models import Group, Principal
+from prometheus_client import Counter, Histogram
+from rest_framework import status
+
+from api.models import User
+from .unexpected_status_code_from_it import UnexpectedStatusCodeFromITError
+
+# Constants or global variables.
+LOGGER = logging.getLogger(__name__)
+TYPE_SERVICE_ACCOUNT = "service-account"
+
+# IT path to fetch the service accounts.
+IT_PATH_GET_SERVICE_ACCOUNTS = "/service_accounts/v1"
+
+# Set up the metrics for the IT calls.
+it_request_all_service_accounts_time_tracking = Histogram(
+    "it_request_all_service_accounts_processing_seconds",
+    "Time spent processing requests from IT when requesting all the service accounts of a tenant",
+)
+
+it_request_status_count = Counter(
+    "it_request_status_total",
+    "Number of requests from IT to BOP and resulting status",
+    ["method", "status"],
+)
+
+
+class ITService:
+    """A class to handle interactions with the IT service."""
+
+    def __init__(self):
+        """Establish IT connection information."""
+        self.host = settings.IT_SERVICE_HOST
+        self.base_path = settings.IT_SERVICE_BASE_PATH
+        self.port = settings.IT_SERVICE_PORT
+        self.protocol = settings.IT_SERVICE_PROTOCOL_SCHEME
+
+    @it_request_all_service_accounts_time_tracking.time()
+    def request_service_accounts(self, bearer_token: str) -> list[dict]:
+        """Request the service accounts for a tenant and returns the entire list that IT has."""
+        received_service_accounts: list[dict] = []
+        # Prepare the URL to fetch the service accounts.
+        url = f"{self.protocol}://{self.host}:{self.port}{self.base_path}{IT_PATH_GET_SERVICE_ACCOUNTS}"
+
+        # Attempt fetching all the service accounts for the tenant.
+        try:
+            offset = 0
+            limit = 100
+
+            # If the offset is zero, that means that we need to call the service at least once to get the first
+            # service accounts. If it equals the limit, that means that there are more pages to fetch.
+            while offset == 0 or offset == limit:
+                response = requests.get(
+                    url=url,
+                    headers={"Authorization": f"Bearer {bearer_token}"},
+                    params={"first": offset, "max": limit},
+                )
```

Comment:
By default, I think the requests package does not set a timeout on the http calls.  If the call hangs, you might start eating up the threads in the thread pool.

https://requests.readthedocs.io/en/latest/user/quickstart/#timeouts
https://requests.readthedocs.io/en/latest/user/advanced/#timeouts

Reply:
Done!